File: python/lightning_sdk/sandbox/network_policy.py
"""Sandbox egress firewall policy (maps to ``V1NetworkPolicy``).

A single :class:`NetworkPolicy` models every egress mode (mirrors Vercel's
``NetworkPolicy``):

  - ``NetworkPolicy("allow-all")`` — open egress (the default)
  - ``NetworkPolicy("deny-all")`` — block all egress
  - ``NetworkPolicy(allow_cidrs=[...])`` — CIDR allowlist (``default-deny`` for
    everything else)

For convenience, :meth:`~lightning_sdk.sandbox.sandbox.Sandbox.create` also
accepts the bare ``"allow-all"`` / ``"deny-all"`` strings. Policies are set at
create time only; restored snapshots inherit the source policy unless
overridden. Reading :attr:`~lightning_sdk.sandbox.base.SandboxInstance.network_policy`
always returns a :class:`NetworkPolicy`.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Literal, TypeAlias

from lightning_sdk.lightning_cloud.openapi.models import V1NetworkPolicy

NetworkPolicyMode: TypeAlias = Literal["allow-all", "deny-all", "default-deny"]
#: Bare-string shorthands accepted at create time.
NetworkPolicyShorthand: TypeAlias = Literal["allow-all", "deny-all"]


@dataclass(frozen=True)
class NetworkPolicy:
    """Sandbox egress firewall policy.

    ``mode`` is one of ``"allow-all"`` (open egress), ``"deny-all"`` (block all),
    or ``"default-deny"`` (deny everything except ``allow_cidrs``). Passing
    ``allow_cidrs`` without an explicit ``mode`` implies ``"default-deny"``.
    """

    mode: NetworkPolicyMode = "allow-all"
    allow_cidrs: list[str] = field(default_factory=list)

    def __post_init__(self) -> None:
        """Infer ``default-deny`` when CIDRs are supplied without an explicit mode."""
        if self.allow_cidrs and self.mode == "allow-all":
            object.__setattr__(self, "mode", "default-deny")

    def to_v1(self) -> V1NetworkPolicy:
        return V1NetworkPolicy(
            mode=self.mode,
            allowed_cidrs=list(self.allow_cidrs),
        )


NetworkPolicyInput: TypeAlias = NetworkPolicyShorthand | NetworkPolicy | V1NetworkPolicy | None


def to_v1_network_policy(
    policy: NetworkPolicyInput,
) -> V1NetworkPolicy | None:
    """Convert SDK policy values (or pass through ``V1NetworkPolicy``)."""
    if policy is None:
        return None
    if isinstance(policy, V1NetworkPolicy):
        return policy
    if isinstance(policy, str):
        return V1NetworkPolicy(mode=policy)
    if isinstance(policy, NetworkPolicy):
        return policy.to_v1()
    raise TypeError(
        f"network_policy must be 'allow-all', 'deny-all', NetworkPolicy, or V1NetworkPolicy, got {type(policy)!r}",
    )


def from_v1_network_policy(
    policy: V1NetworkPolicy | None,
) -> NetworkPolicy:
    """Convert a backend ``V1NetworkPolicy`` into a :class:`NetworkPolicy`.

    Inverse of :func:`to_v1_network_policy`. Open egress (no policy stored) and
    an explicit ``allow-all`` both normalize to ``NetworkPolicy("allow-all")``,
    matching the create-time default, so callers always get a single type and
    never see ``None`` or the generated ``V1NetworkPolicy``.
    """
    if policy is None:
        return NetworkPolicy(mode="allow-all")
    mode = getattr(policy, "mode", None) or "allow-all"
    cidrs = list(getattr(policy, "allowed_cidrs", None) or [])
    if mode not in ("allow-all", "deny-all", "default-deny"):
        mode = "default-deny" if cidrs else "allow-all"
    return NetworkPolicy(mode=mode, allow_cidrs=cidrs)

File: js/src/network-policy.ts
import type { V1NetworkPolicy } from "./lightning_cloud/openapi/data-contracts.js";

/**
 * Egress policy mode. Mirrors the Python SDK's `NetworkPolicy`:
 *
 * - `"allow-all"` — open egress (the default)
 * - `"deny-all"` — block all egress
 * - `"default-deny"` — deny everything except {@link NetworkPolicy.allowCidrs}
 */
export type NetworkPolicyMode = "allow-all" | "deny-all" | "default-deny";

/** Bare-string shorthands accepted at create time. */
export type NetworkPolicyShorthand = "allow-all" | "deny-all";

/** Anything {@link Sandbox.create} accepts for `networkPolicy`. */
export type NetworkPolicyInput =
  | NetworkPolicyShorthand
  | NetworkPolicy
  | V1NetworkPolicy;

/**
 * Sandbox egress firewall policy, set at create time only (no runtime
 * mutation). A single `NetworkPolicy` models every mode:
 *
 * ```ts
 * import { NetworkPolicy, Sandbox } from "@lightning-ai/sandbox";
 *
 * const locked = await Sandbox.create({
 *   name: "egress-test",
 *   instanceType: "cpu-1",
 *   networkPolicy: new NetworkPolicy({ allowCidrs: ["10.0.0.0/8"] }),
 * });
 * console.log(locked.networkPolicy.mode); // "default-deny"
 * ```
 *
 * Passing `allowCidrs` without an explicit `mode` implies `"default-deny"`.
 */
export class NetworkPolicy {
  readonly mode: NetworkPolicyMode;
  readonly allowCidrs: string[];

  constructor(opts: { mode?: NetworkPolicyMode; allowCidrs?: string[] } = {}) {
    const allowCidrs = opts.allowCidrs ?? [];
    let mode = opts.mode ?? "allow-all";
    // Infer default-deny when CIDRs are supplied without an explicit mode.
    if (allowCidrs.length > 0 && mode === "allow-all") {
      mode = "default-deny";
    }
    this.mode = mode;
    this.allowCidrs = allowCidrs;
  }

  toV1(): V1NetworkPolicy {
    return { mode: this.mode, allowedCidrs: [...this.allowCidrs] };
  }
}

/** Convert SDK policy values (or pass through a raw `V1NetworkPolicy`). */
export function toV1NetworkPolicy(
  policy: NetworkPolicyInput | undefined,
): V1NetworkPolicy | undefined {
  if (policy === undefined) return undefined;
  if (typeof policy === "string") return { mode: policy };
  if (policy instanceof NetworkPolicy) return policy.toV1();
  return policy;
}

/**
 * Convert a backend `V1NetworkPolicy` into a {@link NetworkPolicy}. Open egress
 * (no stored policy) and an explicit `allow-all` both normalize to
 * `new NetworkPolicy({ mode: "allow-all" })`, so callers never get `undefined`
 * or a raw `V1NetworkPolicy` back.
 */
export function fromV1NetworkPolicy(
  policy: V1NetworkPolicy | undefined,
): NetworkPolicy {
  if (!policy) return new NetworkPolicy({ mode: "allow-all" });
  const cidrs = policy.allowedCidrs ?? [];
  let mode = policy.mode ?? "allow-all";
  if (mode !== "allow-all" && mode !== "deny-all" && mode !== "default-deny") {
    mode = cidrs.length > 0 ? "default-deny" : "allow-all";
  }
  return new NetworkPolicy({ mode: mode as NetworkPolicyMode, allowCidrs: cidrs });
}
